components/brave_news/browser/peeking_card.cc

```diff
@@ -0,0 +1,214 @@
+
+      // Boost enabled publishers
+    } else if (subscriptions.enabled_publishers().contains(
+                   article->publisher_id)) {
+      score += kPublisherBoost;
+
+      // Boost enabled channels.
+    } else if (std::ranges::any_of(article->channels, [&subscribed_channels](
+                                                          const auto& channel) {
+                 return base::Contains(subscribed_channels, channel);
+               })) {
+      score += kChannelBoost;
+    }
+
+    if (base::Contains(top_story_urls, article->url.spec())) {
+      score *= kTopStoryMultiplier;
+    }
+
+    // Article has no score, so we can't do anything with it.
+    if (score == 0) {
+      continue;
+    }
+
+    // Apply a boost to recent articles.
+    auto elapsed = now - article->publish_time;
+    if (elapsed <= base::Hours(1)) {
+      score *= kOneHourMultiplier;
+    } else if (elapsed <= base::Days(1)) {
+      score *= kOneDayMultiplier;
+    } else {
+      // Decay for a week - after a week there probably isn't much difference.
+      // (a half life of 1.1 days does nicely here)
+      score *= std::max(0.1, pow(0.5, elapsed.InDays() / 1.1));
+    }
+
+    candidates.emplace_back(i, score);
+  }
+
+  base::Time::Exploded exploded;
+  now.LocalExplode(&exploded);
+  for (auto& [index, score] : candidates) {
+    const auto& article = get_article(index);
+    // In the morning, weight news higher
+    if (6 <= exploded.hour && exploded.hour < 10 &&
+        base::Contains(article->channels, kTopNewsChannel)) {
+      score += kMorningNewsBoost;
+
+      // In the evening, weight entertainment higher
+    } else if (17 <= exploded.hour && exploded.hour <= 22 &&
+               base::Contains(article->channels, kEntertainmentChannel)) {
+      score += kEveningEntertainmentBoost;
+    }
+  }
+
+  base::ranges::sort(candidates, [get_article](const auto& a, const auto& b) {
+    const auto& [a_index, a_score] = a;
+    const auto& [b_index, b_score] = b;
+    if (a_score != b_score) {
+      return a_score > b_score;
+    }
+
+    const auto& a_article = get_article(a_index);
+    const auto& b_article = get_article(b_index);
+
+    return a_article->publish_time > b_article->publish_time;
+  });
+
+  std::vector<ItemScore> final_candidates;
+  std::map<std::string, size_t> seen_channels;
+
+  // Make sure we can pick enough options to have at least |maxCandidates|
+  // options in out finalCandidates.
+  const auto channel_limit = following_count < kMaxPeekingCardCandidates
+                                 ? (max_candidates / following_count)
+                                 : 1;
+  for (auto& [index, score] : candidates) {
+    if (final_candidates.size() > max_candidates) {
+      break;
+    }
+
+    const auto& article = get_article(index);
+
+    // We don't want any channel to dominate our final candidates list.
+    if (std::ranges::any_of(article->channels, [&seen_channels, channel_limit](
```

Comment:
This will cause `Top Sources` and other umbrellas channels ('Entertainment') to be counted in and limit more specific channels, example: if two publishers are `Top Sources` we would select only from one, similarly we could only have either a Music candidate or a Film and TV one, because they both belong to Entertainment. 

Since we enforce a less specific -> more specific order in the backend, we could just use `article->channels.last` (ie the most specific) here. As long as the client has maintained that order up to here.